lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -874,26 +884,31 @@ sub _add_job {
     # add parents
     for my $parent ($job->parents->all) {
         my ($parent_job, $dependency_type) = ($parent->parent, $parent->dependency);
-        my $parent_job_id = _add_job($visited, $nodes, $edges, $cluster, $cluster_by_job, $parent_job) or next;
+        my $parent_job_id
+          = _add_job($visited, $nodes, $edges, $cluster, $cluster_by_job, $parent_job, $as_child_of, $preferred_depth)
+          or next;
         _add_dependency_to_graph($edges, $cluster, $cluster_by_job, $parent_job_id, $job_id, $dependency_type);
         _add_dependency_to_node(\%node, $parent_job, $dependency_type);
     }
 
     # add children
     for my $child ($job->children->all) {
-        _add_job($visited, $nodes, $edges, $cluster, $cluster_by_job, $child->child);
+        # add chained deps only if we're still on the preferred depth to avoid dragging too many jobs into the tree
+        next
+          if ($ancestors //= $job->ancestors) != $preferred_depth
+          && $child->dependency != OpenQA::JobDependencies::Constants::PARALLEL;
+        _add_job($visited, $nodes, $edges, $cluster, $cluster_by_job, $child->child, $job_id, $preferred_depth);
     }
 
     return $job_id;
 }
 
-sub dependencies {
-    my ($self) = @_;
+sub dependencies ($self) {
 
```

Comment:
to reduce risk of deployment I suggest to split out introducing signatures into a separate commit and even separate PR that we can merge first